audb/core/stream.py

```diff
@@ -0,0 +1,596 @@
+        _update_path(
+            self,
+            self.root,
+            self._full_path,
+            self._format,
+            self._num_workers,
+            self._verbose,
+        )
+
+        return df
+
+    @staticmethod
+    def _cleanup_database(db: audformat.Database, table: str):
+        r"""Remove parts of database, not used by table.
+
+        Args:
+            db: database object
+            table: table ID
+
+        """
+        tables = [table]
+        misc_tables = _misc_tables_used_in_table(db[table])
+
+        # Remove non-requested table
+        db.drop_tables([table for table in list(db.tables) if table not in tables])
+
+        # Remove unused schemes
+        used_schemes = []
+        for table in misc_tables + tables:
+            for column_id, column in db[table].columns.items():
+                if column.scheme_id is not None:
+                    used_schemes.append(column.scheme_id)
+        for scheme in list(db.schemes):
+            if scheme not in used_schemes:
+                del db.schemes[scheme]
+
+        # Remove misc tables not required by the schemes of table
+        db.drop_tables(
+            [
+                misc_table
+                for misc_table in list(db.misc_tables)
+                if misc_table not in misc_tables + tables
+            ]
+        )
+
+        # Remove unused splits
+        used_splits = [
+            db[table].split_id for table in list(db) if db[table].split_id is not None
+        ]
+        for split in list(db.splits):
+            if split not in used_splits:
+                del db.splits[split]
+
+    def _get_batch(self) -> pd.DataFrame:
+        r"""Read table batch.
+
+        Returns:
+            dataframe
+
+        """
+        if self._shuffle:
+            buffer_read_length = self._batch_size
+            df1 = pd.DataFrame()
+            if len(self._buffer) < self._batch_size:
+                if len(self._buffer) > 0:
+                    # Empty current buffer,
+                    # before refilling
+                    df1 = self._buffer
+                    buffer_read_length = self._batch_size - len(self._buffer)
+                self._buffer = self._read_dataframe()
+                # Shuffle data
+                self._buffer = self._buffer.sample(frac=1)
+
+            df2 = self._buffer.iloc[:buffer_read_length, :]
+            self._buffer.drop(index=df2.index, inplace=True)
+            df = pd.concat([df1, df2])
+
+        else:
+            df = self._read_dataframe()
+
+        if len(df) == 0:
+            raise StopIteration
+
+        return df
+
+    def _initialize_stream(self) -> typing.Iterable:
+        r"""Create table iterator object.
+
+        This method needs to be implemented
+        for the table file types
+        in the classes,
+        that inherit from :class:`audb.DatabaseIterator`.
+
+        Returns:
+            table iterator
+
+        """
+        return None  # pragma: nocover
```

Comment:
Is this kind of an  `interface` in the java sense of the word interface?
You mention that this  "method needs to be implemented".

More commonly you would probably have it as an `@abstractmethod` decorator from the abc modules. Some people would also more simply raise a `NotImplementedError`.

My knowdledge is not fully up to date and I have not read the docs about [protocols](https://typing.readthedocs.io/en/latest/spec/protocol.html#protocols) - but I am interested in doing so: interestingly the documentation on protocols discusses iterators.

Reply:
I converted `DatabaseIterator` into an abstract class (using `metaclass=abc.ABCMeta`) and added the `@abc.abstractmethod` decorator to `_initialize_stream()`.

Reply:
Looks good.  `_postprocess_batch`  is probably also correct - there is a default implmentation that is simply overridden. Resolving the conversation.